src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,161 @@
+import duke.exceptions.InvalidDateTimeException;
+import duke.exceptions.InvalidInputException;
+import duke.exceptions.InvalidParametersException;
+import duke.exceptions.InvalidTaskException;
+
+/**
+ * This class represents the parser used to receive and handle user inputs to the UI.
+ */
+public class Parser {
+    /**
+     * Parses command from array.
+     *
+     * @param arr The array to parse.
+     * @return A correct command enum assigned to the array.
+     * @throws InvalidTaskException When the command does not match any of the enums.
+     */
+    public static UI.Command getCommand(String[] arr) throws InvalidTaskException {
+        switch (arr[0]) {
+        case "bye":
+            return UI.Command.BYE;
+        case "todo":
+            return UI.Command.TODO;
+        case "event":
+            return UI.Command.EVENT;
+        case "deadline":
+            return UI.Command.DEADLINE;
+        case "list":
+            return UI.Command.LIST;
+        case "unmark":
+            return UI.Command.UNMARK;
+        case "mark":
+            return UI.Command.MARK;
+        case "delete":
+            return UI.Command.DELETE;
+        case "find":
+            return UI.Command.FIND;
+        default:
+            throw new InvalidTaskException();
+        }
+    }
+    /**
+     * Generates the index with respect to a zero-indexed task array.
+     *
+     * @param s The index in question.
+     * @param bounds The bounds of the task array.
+     * @return The correct index with respect to the task array.
+     * @throws HistoryIndexException The index is invalid.
+     */
+    public static Integer checkIndexGiven(String s, int bounds) throws HistoryIndexException {
+        Integer parsed = Integer.parseInt(s);
+        if (parsed <= 0 || parsed > bounds) {
+            throw new HistoryIndexException();
+        }
+        return parsed - 1;
+    }
+    /**
+     * Extracts description data from an array.
+     *
+     * @param descriptionArray The array that holds description data to parse.
+     * @return An array representing the extracted data from the description array.
+     * @throws InvalidInputException The given input has something wrong (Parameter wise or command wise).
+     */
+    public static String[] extractDescriptionData(String... descriptionArray) throws
+            InvalidInputException {
+        String[] ret = new String[3];
+        String taskDesc;
+        switch(descriptionArray[0]) {
+        case "todo":
+            taskDesc = String.join(" ",
+                Arrays.copyOfRange(descriptionArray, 1, descriptionArray.length));
+            ret[0] = taskDesc;
+            break;
+        case "find":
+            String searchDesc = String.join(" ",
+                Arrays.copyOfRange(descriptionArray, 1, descriptionArray.length));
+            ret[0] = searchDesc;
+            break;
+        case "event":
+            Integer startIdx = -1;
+            Integer endIdx = -1;
```

Comment:
Perhaps naming the variables more intuitively can reduce the chances of future confusion!
```suggestion
            Integer startIndex = -1;
            Integer endIndex = -1;
```